### PR TITLE
Display addition information about observations

### PR DESCRIPTION
**Pull request by Matt-Guest**
### What

Display `data markings` or any other additional fields which are related to the observation.

If UserNotes are provided by the dataset api display them on the excel file (Work in progress)

### How to review

Review the code, run tests

### Who can review

@CarlHembrough or @daiLlew

### CODE REVIEW

File: src/main/java/dp/xlsx/DatasetFormatter.java [hunk 1]
```diff
@@ -84,20 +78,32 @@ void format() {
             sheet.setColumnWidth(columnWidth.getKey(),
                     (columnWidth.getValue() + DIMENSION_WIDTH_PADDING_CHARS) * EXCEL_CHARS_TO_WIDTH_FACTOR);
         }
+
+        if (datasetMetadata.getUserNotes() == null) return;
```
Comment (daiLlew) [inline src/main/java/dp/xlsx/DatasetFormatter.java]: picky I know but can you put the brackets in for this `if ` thanks.

File: src/main/java/dp/xlsx/DatasetFormatter.java [hunk 2]
```diff
@@ -197,19 +219,40 @@ private void addMetadata() {
...
+    private void addNote(final String value) {
+        rowOffset++;
+        Row row = sheet.createRow(rowOffset);
+        row.createCell(0); //Blank cell
+        Cell cell = row.createCell(1);
+        cell.setCellValue(value);
+        cell.setCellStyle(workBookStyles.getNoteStyle());
+    }
+
+    private void AddUserNotes(UserNotes notes) {
```
Comment (daiLlew) [inline src/main/java/dp/xlsx/DatasetFormatter.java]: Again picky can you make this method name start with lower case? This aint golang bro :-p

Comment (Matt-Guest) [reply src/main/java/dp/xlsx/DatasetFormatter.java]: Renamed the method 😞

File: src/main/java/dp/xlsx/DatasetFormatter.java (hunk 1 above)
Comment (Matt-Guest) [reply src/main/java/dp/xlsx/DatasetFormatter.java]: I inverted the if statement and moved for loop into it. 👍

File: src/main/java/dp/api/dataset/models/Metadata.java [hunk 3]
```diff
@@ -83,6 +83,9 @@
     @JsonProperty("uri")
     private String uri;
 
+    @JsonProperty("user_notes")
+    private UserNotes[] userNotes;
```
Comment (CarlHembrough) [inline src/main/java/dp/api/dataset/models/Metadata.java]: * usageNotes?

File: src/main/java/dp/xlsx/DatasetFormatter.java [hunk 4]
```diff
@@ -84,20 +78,33 @@ void format() {
...
+                Cell data = row.createCell(columnOffset);
+                data.setCellValue(additionalValue);
+                setCellValueByType(data, additionalValue);
+                columnOffset++;
+            }
 
-            columnOffset++;
+            if (observationValue != null && observationValue.length() > widestDataColumn)
+                widestDataColumn = observationValue.length();
         }
```
Comment (CarlHembrough) [inline src/main/java/dp/xlsx/DatasetFormatter.java]: Do we know if any additional values will be longer than the observations?

File: src/main/java/dp/xlsx/DatasetFormatter.java [hunk 5]
```diff
@@ -160,16 +167,32 @@ private void addHeaderRow(Collection<String> timeLabels) {
...
         // write time labels across the title row
         for (String timeLabel : timeLabels) {
             Cell cell = headerRow.createCell(columnOffset);
-            cell.setCellStyle(valueRightAlign);
+            cell.setCellStyle(workBookStyles.getValueRightAlignStyle());
             cell.setCellValue(timeLabel);
             columnOffset++;
+
+            for (String additionalHeader : additionalHeaders) {
+                addHeaderCell(columnOffset, headerRow, additionalHeader + " (" + timeLabel + ")");
```
Comment (CarlHembrough) [inline src/main/java/dp/xlsx/DatasetFormatter.java]: I can only imagine that this cell will be the widest in the column. If so we will have to make sure the column width is as wide as the label

File: src/main/java/dp/xlsx/DatasetFormatter.java [hunk 6]
```diff
@@ -160,16 +167,32 @@ private void addHeaderRow(Collection<String> timeLabels) {
...
+                columnOffset++;
+            }
         }
     }
 
 
+    private void setCellValueByType(Cell cell, String value) {
+
+        if (value.chars().allMatch(Character::isDigit)) {
```
Comment (CarlHembrough) [inline src/main/java/dp/xlsx/DatasetFormatter.java]: will have to check this doesn't add too much overhead compared to the previous implementation

File: src/main/java/dp/api/dataset/models/Metadata.java (hunk 3 above)
Comment (Matt-Guest) [reply src/main/java/dp/api/dataset/models/Metadata.java]: done

File: src/main/java/dp/xlsx/DatasetFormatter.java (hunk 5 above)
Comment (Matt-Guest) [reply src/main/java/dp/xlsx/DatasetFormatter.java]: I raised the column factor size

**Pull request merged**